internal/plugin/grpc_provider.go

```diff
@@ -102,17 +147,33 @@ func (p *GRPCProvider) GetProviderSchema() (resp providers.GetProviderSchemaResp
 	// this for compatibility, but recent providers all set the max message
 	// size much higher on the server side, which is the supported method for
 	// determining payload size.
-	const maxRecvSize = 64 << 20
-	protoResp, err := p.client.GetSchema(p.ctx, new(proto.GetProviderSchema_Request), grpc.MaxRecvMsgSizeCallOption{MaxRecvMsgSize: maxRecvSize})
-	if err != nil {
-		resp.Diagnostics = resp.Diagnostics.Append(grpcErr(err))
-		return resp
-	}
+	if protoResp == nil {
+		const maxRecvSize = 64 << 20
+		var err error
+		protoResp, err = p.client.GetSchema(p.ctx, new(proto.GetProviderSchema_Request), grpc.MaxRecvMsgSizeCallOption{MaxRecvMsgSize: maxRecvSize})
+		if err != nil {
+			resp.Diagnostics = resp.Diagnostics.Append(grpcErr(err))
+			return resp
+		}
 
-	resp.Diagnostics = resp.Diagnostics.Append(convert.ProtoToDiagnostics(protoResp.Diagnostics))
+		// UPBOUND CHANGE: cache provider schema in memory and on disk.
+		externalProvider := p.PluginClient.ReattachConfig() != nil
+		if !resp.Diagnostics.HasErrors() && cacheDir != "" && (resp.ServerCapabilities.GetProviderSchemaOptional || externalProvider) {
```

Comment:
Assuming the above comment holds:
```suggestion
		if !resp.Diagnostics.HasErrors() && cacheDir != "" && p.Addr.IsZero() {
```

The idea is if this is a long running plugin process, then in order to be able to cache the schema, `GetProviderSchema` must have been called once, hence we drop the check for whether the provider does any initializations with this call, i.e., we drop the check on `resp.ServerCapabilities.GetProviderSchemaOptional`.

I think for the official providers use case, this is not important because all the generated provider configurations will depend on the same native provider version. But in the general case (e.g., if we would like to contribute back to the upstream repo), we had better consider the native provider version and that version is not part of the Terraform provider address:
https://github.com/hashicorp/terraform-registry-address/blob/ebfd28edbe9c58c616349106dc890e014ff25f6a/provider.go#L16

Reply:
The reattachconfig includes the pid. Isn't that enough to make the cache file unique enough?